src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -51,23 +60,35 @@ public class AsyncSemaphore : IDisposable
         /// <param name="initialCount">The initial number of requests for the semaphore that can be granted concurrently.</param>
         public AsyncSemaphore(int initialCount)
         {
-            this.semaphore = new SemaphoreSlim(initialCount);
+            this.CurrentCount = initialCount;
             this.uncontestedReleaser = Task.FromResult(new Releaser(this));
-
-            this.canceledReleaser = Task.FromCanceled<Releaser>(new CancellationToken(canceled: true));
         }
 
         /// <summary>
         /// Gets the number of openings that remain in the semaphore.
         /// </summary>
-        public int CurrentCount => this.semaphore.CurrentCount;
+        public int CurrentCount { get; private set; }
```

Comment:
Are you ok with this having an arbitrary value after disposal? Otherwise you should reset it during `Dispose` and also check for disposal in `Release` to not increment it again after disposal.

Reply:
It's a good question. I need to take some time to think through and document in the api doc comments what the actual behavior is expected to be at and after disposal with regard to current semaphore users, current waiters, and future requests. Then I'll circle back with a better answer for you here.

Reply:
I've added docs to describe the disposal behavior. Previous to this change we inherited SemaphoreSlim's CurrentCount which retained its value after disposal. So I've opted to do the same to avoid an unnecessary observable behavior change.